EntryStore - Serialization Exception on IMap.put

<!--
Thanks for reporting your issue. Please share with us the following information, to help us resolve your issue quickly and efficiently.
-->

**Describe the bug**

When using the new EntryStore API, and issuing a "PUT" into the IMap for an item which is NOT in the IMap but does exist in the backing store (database), the DefaultRecordStore will fetch the value (which is a MetadataAwareValue) from the store and attempt to return it without unraveling the value.  This causes HZ to throw the following error:

Exception in thread "main" com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'com.hazelcast.map.EntryLoader$MetadataAwareValue'

**Expected behavior**

The store should be able to return a value which has not been loaded yet into the IMap and the value should be returned as the result of the put operation, without the serialization exception.
 

**To Reproduce**

Steps to reproduce the behavior:

1. Using the sample at https://github.com/hazelcast/hazelcast-code-samples/tree/master/distributed-map/entrystore
2. Modify the DummyDatabaseConnection.java replacing the "lookup" method with the following:

```
    public Record lookup(String key) {
        if ( key.equalsIgnoreCase("one-minute-key") )
        {
            storage.put("one-minute-key", new Record("one-minute-key","someval",System.currentTimeMillis() + 60000));
        }
        return storage.get(key);
    }
```
3. Execute the sample's main class "EntryStoreSample"


**Additional context**

The issue appears to be related to how the DefaultRecordStore handles the putInternal call

`        Object oldValue = record == null ? (loadFromStore ? this.mapDataStore.load(key) : null) : record.getValue();`

It seems the _mapDataStore_ should be checked to see if it's an EntryStore or the returned value should be checked for being an instance of MetadataAwareValue, in which case the value should be retrieved from the instance.
 
```
    protected Object putInternal(Data key, Object newValue, long ttl, long maxIdle, @Nullable UUID transactionId, boolean loadFromStore, boolean countAsAccess) {
        this.checkIfLoaded();
        long now = getNow();
        this.markRecordStoreExpirable(ttl, maxIdle);
        Record record = this.getRecordOrNull(key, now, false);
        Object oldValue = record == null ? (loadFromStore ? this.mapDataStore.load(key) : null) : record.getValue();
        newValue = this.mapServiceContext.interceptPut(this.interceptorRegistry, oldValue, newValue);
        this.onStore(record);
        if (record == null) {
            this.putNewRecord(key, oldValue, newValue, ttl, maxIdle, now, transactionId);
        } else {
            this.updateRecord(key, record, oldValue, newValue, now, countAsAccess, ttl, maxIdle, true, transactionId, false);
        }
        return oldValue;
    }
```


**Detailed StackTrace:**

```
Exception in thread "main" com.hazelcast.nio.serialization.HazelcastSerializationException: Failed to serialize 'com.hazelcast.map.EntryLoader$MetadataAwareValue'
                at com.hazelcast.internal.serialization.impl.SerializationUtil.handleSerializeException(SerializationUtil.java:115)
                at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:173)
                at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:149)
                at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:134)
                at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toData(AbstractSerializationService.java:122)
                at com.hazelcast.internal.serialization.impl.SerializationServiceV1.toData(SerializationServiceV1.java:137)
                at com.hazelcast.map.impl.MapServiceContextImpl.toData(MapServiceContextImpl.java:548)
                at com.hazelcast.map.impl.operation.PutOperation.runInternal(PutOperation.java:36)
                at com.hazelcast.map.impl.operation.MapOperation.run(MapOperation.java:112)
                at com.hazelcast.spi.impl.operationservice.Operation.call(Operation.java:184)
                at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.call(OperationRunnerImpl.java:227)
                at com.hazelcast.spi.impl.operationservice.impl.OperationRunnerImpl.run(OperationRunnerImpl.java:216)
                at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:160)
                at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.process(OperationThread.java:138)
                at com.hazelcast.spi.impl.operationexecutor.impl.OperationThread.run(OperationThread.java:123)
Caused by: com.hazelcast.nio.serialization.HazelcastSerializationException: There is no suitable serializer for class com.hazelcast.map.EntryLoader$MetadataAwareValue
                at com.hazelcast.internal.serialization.impl.AbstractSerializationService.serializerFor(AbstractSerializationService.java:507)
                at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toBytes(AbstractSerializationService.java:162)
```

Hazelcast: 4.0.2


<!--
Add any other context about the problem here.

Common details that we're often interested in:
- Detailed description of the steps to reproduce your issue
- Logs and stack traces, if available
- Hazelcast version that you use (e.g. 3.4, also specify whether it is a minor release or the latest snapshot)
- If available, integration module versions (e.g. Tomcat, Jetty, Spring, Hibernate). Also, include their detailed configuration information such as web.xml, Hibernate configuration and `context.xml` for Spring
- Cluster size, i.e. the number of Hazelcast cluster members
- Number of the clients
- Version of Java. It is also helpful to mention the JVM parameters
- Operating system. If it is Linux, kernel version is helpful
- Unit test with the `hazelcast.xml` file. If you could include a unit test which reproduces your issue, we would be grateful
-->

--- a/hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java
+++ b/hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java
@@ -363,6 +363,31 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore {
         return lockStore != null ? lockStore.getOwnerInfo(key) : null;
     }
 
+    /**
+     * Loads value of key. If necessary loads it by
+     * extracting from {@link MetadataAwareValue}
+     *
+     * @return loaded value from map-store,
+     * when no value found returns null
+     */
+    private Object loadValueOf(Data key) {
+        Object value = mapDataStore.load(key);
+        if (value == null) {
+            return null;
+        }
+
+        if (mapDataStore.isWithExpirationTime()) {
+            MetadataAwareValue loaderEntry = (MetadataAwareValue) value;
+            long proposedTtl = expirationTimeToTtl(loaderEntry.getExpirationTime());
+            if (proposedTtl <= 0) {
+                return null;
+            }
+            value = loaderEntry.getValue();
+        }
+
+        return value;
+    }
+
     @Override
     public Record loadRecordOrNull(Data key, boolean backup, Address callerAddress) {
         Object value = mapDataStore.load(key);
@@ -503,7 +528,7 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore {
         Record record = getRecordOrNull(key, now, false);
         Object oldValue;
         if (record == null) {
-            oldValue = mapDataStore.load(key);
+            oldValue = loadValueOf(key);
             if (oldValue != null && persistenceEnabledFor(provenance)) {
                 mapDataStore.remove(key, now, transactionId);
             }
@@ -522,7 +547,7 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore {
         Object oldValue;
         boolean removed = false;
         if (record == null) {
-            oldValue = mapDataStore.load(key);
+            oldValue = loadValueOf(key);
             if (oldValue == null) {
                 return false;
             }
@@ -718,7 +743,7 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore {
     public boolean setTtl(Data key, long ttl, boolean backup) {
         long now = getNow();
         Record record = getRecordOrNull(key, now, false);
-        Object existingValue = record == null ? mapDataStore.load(key) : record.getValue();
+        Object existingValue = record == null ? loadValueOf(key) : record.getValue();
         if (existingValue == null) {
             return false;
         }
@@ -758,7 +783,7 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore {
 
         Record record = getRecordOrNull(key, now, false);
         Object oldValue = record == null
-                ? (loadFromStore ? mapDataStore.load(key) : null) : record.getValue();
+                ? (loadFromStore ? loadValueOf(key) : null) : record.getValue();
         newValue = mapServiceContext.interceptPut(interceptorRegistry, oldValue, newValue);
         onStore(record);
 
@@ -843,7 +868,7 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore {
         Record record = getRecordOrNull(key, now, false);
         Object oldValue;
         if (record == null) {
-            oldValue = mapDataStore.load(key);
+            oldValue = loadValueOf(key);
         } else {
             oldValue = record.getValue();
         }
@@ -869,7 +894,7 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore {
         Record record = getRecordOrNull(key, now, false);
         Object current;
         if (record == null) {
-            current = mapDataStore.load(key);
+            current = loadValueOf(key);
         } else {
             current = record.getValue();
         }
@@ -1015,13 +1040,8 @@ public class DefaultRecordStore extends AbstractEvictableRecordStore {
         Record record = getRecordOrNull(key, now, false);
         Object oldValue;
         if (record == null) {
-            oldValue = mapDataStore.load(key);
-            if (oldValue != null) {
-                record = createRecord(key, oldValue, UNSET, UNSET, now);
-                storage.put(key, record);
-                mutationObserver.onPutRecord(key, record, null, false);
-                mapEventPublisher.publishEvent(callerAddress, name, EntryEventType.LOADED, key, null, oldValue);
-            }
+            Record loadedRecord = loadRecordOrNull(key, false, callerAddress);
+            oldValue = loadedRecord != null ? loadedRecord.getValue() : null;
         } else {
             accessRecord(record, now);
             oldValue = record.getValue();

--- a/hazelcast/src/test/java/com/hazelcast/map/EvictionTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/map/EvictionTest.java
@@ -200,9 +200,9 @@ public class EvictionTest extends HazelcastTestSupport {
         IMap<Integer, Employee> map = node.getMap(mapName);
         map.addIndex(type, "city");
 
-        int entries = 5;
+        int entryCount = 5;
         Map<Integer, Long> lastAccessTimes = new HashMap<>();
-        for (int i = 0; i < entries; ++i) {
+        for (int i = 0; i < entryCount; ++i) {
             String cityName = i % 2 == 0 ? "cityname" : null;
 
             Employee emp = new Employee(i, "name" + i, cityName, 0, true, i);
@@ -221,12 +221,15 @@ public class EvictionTest extends HazelcastTestSupport {
         Predicate predicateCityNull = entryObject.get("city").isNull();
         Collection<Employee> valuesNullCity = map.values(predicateCityNull);
         Collection<Employee> valuesNotNullCity = map.values(Predicates.equal("city", "cityname"));
-        assertEquals(entries, valuesNullCity.size() + valuesNotNullCity.size());
+        assertEquals(entryCount, valuesNullCity.size() + valuesNotNullCity.size());
         // check that evaluating the predicate updated the last access time of the returned records
-        for (int i = 0; i < entries; ++i) {
+        for (int i = 0; i < entryCount; ++i) {
             EntryView view = map.getEntryView(i);
             assertNotNull(view);
-            assertTrue(view.getLastAccessTime() > lastAccessTimes.get(i));
+            long lastAccessTime = view.getLastAccessTime();
+            long prevLastAccessTime = lastAccessTimes.get(i);
+            assertTrue("lastAccessTime=" + lastAccessTime + ", prevLastAccessTime=" + prevLastAccessTime,
+                    lastAccessTime > prevLastAccessTime);
         }
     }
 

--- a/hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/EntryLoaderSimpleTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/EntryLoaderSimpleTest.java
@@ -56,7 +56,6 @@ import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicBoolean;
 
 import static java.util.Arrays.asList;
 import static org.junit.Assert.assertEquals;
@@ -386,15 +385,7 @@ public class EntryLoaderSimpleTest extends HazelcastTestSupport {
         Config config = mapServiceContext.getNodeEngine().getConfig();
         MapContainer mapContainer = new MapContainer("anyName", config, mapServiceContext);
         Data key = mapServiceContext.toData("key");
-        final AtomicBoolean invoked = new AtomicBoolean();
-        DefaultRecordStore recordStore = new DefaultRecordStore(mapContainer, 0, mock(MapKeyLoader.class), mock(ILogger.class)) {
-            @Override
-            protected long expirationTimeToTtl(long definedExpirationTime) {
-                invoked.set(true);
-                return 0;
-            }
-        };
+        DefaultRecordStore recordStore = new DefaultRecordStore(mapContainer, 0, mock(MapKeyLoader.class), mock(ILogger.class)) ;
         assertNull(recordStore.loadRecordOrNull(key, false, null));
-        assertTrue(invoked.get());
     }
 }

--- a/hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/EntryStoreSimpleTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/map/impl/mapstore/EntryStoreSimpleTest.java
@@ -21,6 +21,7 @@ import com.hazelcast.config.InMemoryFormat;
 import com.hazelcast.config.MapStoreConfig;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.map.IMap;
+import com.hazelcast.map.listener.EntryAddedListener;
 import com.hazelcast.query.Predicates;
 import com.hazelcast.test.HazelcastParallelParametersRunnerFactory;
 import com.hazelcast.test.HazelcastTestSupport;
@@ -36,12 +37,15 @@ import org.junit.runners.Parameterized.Parameters;
 import org.junit.runners.Parameterized.UseParametersRunnerFactory;
 
 import java.util.Collection;
-import java.util.HashMap;
 import java.util.Map;
+import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
 
 import static java.util.Arrays.asList;
+import static org.junit.Assert.assertEquals;
 
 @RunWith(Parameterized.class)
 @UseParametersRunnerFactory(HazelcastParallelParametersRunnerFactory.class)
@@ -93,6 +97,36 @@ public class EntryStoreSimpleTest extends HazelcastTestSupport {
     }
 
     @Test
+    public void testPut_returns_old_value_from_entry_store() {
+        // 1. Insert value into entry-store
+        map.put("key", "value1");
+
+        // 2. Then remove it only from memory, entry still exists in entry-store
+        map.evict("key");
+
+        // 3. Update entry value to a new one
+        String old = map.put("key", "value2");
+
+        // 4. Expect we got correct old value
+        assertEquals("value1", old);
+    }
+
+    @Test
+    public void testReplace_returns_old_value_from_entry_store() {
+        // 1. Insert value into entry-store
+        map.put("key", "value1");
+
+        // 2. Then remove it only from memory, entry still exists in entry-store
+        map.evict("key");
+
+        // 3. Replace entry value with a new one
+        String old = map.replace("key", "value2");
+
+        // 4. Expect we got correct old value
+        assertEquals("value1", old);
+    }
+
+    @Test
     public void testPut_withTtl() {
         map.put("key", "value", 10, TimeUnit.DAYS);
         assertEntryStore("key", "value", 10, TimeUnit.DAYS, 10000);
@@ -119,15 +153,31 @@ public class EntryStoreSimpleTest extends HazelcastTestSupport {
     }
 
     @Test
-    public void testPutAll() {
-        Map<String, String> businessObjects = new HashMap<>();
-        for (int i = 0; i < 100; i++) {
-            businessObjects.put("k" + i, "v" + i);
-        }
+    public void testPutAll_WithoutMapListener() {
+        final int max = 100;
+        final Map<String, String> businessObjects = IntStream.range(0, max).boxed()
+                .collect(Collectors.toMap(i -> "k" + i, i -> "v" + i));
+
         map.putAll(businessObjects);
-        for (int i = 0; i < 100; i++) {
-            assertEntryStore("k" + i, "v" + i);
-        }
+
+        IntStream.range(0, max).forEach(i -> assertEntryStore("k" + i, "v" + i));
+        assertEquals(0, testEntryStore.getLoadCallCount());
+    }
+
+    @Test
+    public void testPutAll_WithMapListener() {
+        final int max = 100;
+        final Map<String, String> businessObjects = IntStream.range(0, max).boxed()
+                .collect(Collectors.toMap(i -> "k" + i, i -> "v" + i));
+
+        final CountDownLatch latch = new CountDownLatch(max);
+        map.addEntryListener((EntryAddedListener) event -> latch.countDown(), true);
+
+        map.putAll(businessObjects);
+
+        IntStream.range(0, max).forEach(i -> assertEntryStore("k" + i, "v" + i));
+        assertEquals(max, testEntryStore.getLoadCallCount());
+        assertOpenEventually(latch);
     }
 
     @Test

